File: frontend/mobile/apple/iOS/My Home Agent/My Home Agent/Config/AppConfiguration.swift
//
//  AppConfiguration.swift
//  My Home Agent
//
//  Build configuration management for development and production environments
//

import Foundation
import OSLog

/// Build environment types
enum BuildEnvironment: String {
    case development
    case staging
    case production
    
    /// Current build environment based on compiler flags
    static var current: BuildEnvironment {
        #if DEBUG
        return .development
        #elseif STAGING
        return .staging
        #else
        return .production
        #endif
    }
}

/// Centralized app configuration based on build environment
struct AppConfiguration {
    
    // MARK: - Environment
    
    /// Current build environment
    static let environment = BuildEnvironment.current
    
    // MARK: - API Configuration
    
    /// Base URL for API server
    static var apiBaseURL: String {
        switch environment {
        case .development:
            return "https://dev-homeservices.hestami-ai.com"
        case .staging:
            return "https://staging-homeservices.hestami-ai.com"
        case .production:
            return "https://homeservices.hestami-ai.com"
        }
    }
    
    /// Static media server host
    static var staticMediaHost: String {
        switch environment {
        case .development:
            return "dev-static.hestami-ai.com"
        case .staging:
            return "staging-static.hestami-ai.com"
        case .production:
            return "static.hestami-ai.com"
        }
    }
    
    /// Static media server port
    static var staticMediaPort: String {
        return "443" // HTTPS port for all environments
    }
    
    // MARK: - Feature Flags
    
    /// Enable verbose logging (debug builds only)
    static var enableVerboseLogging: Bool {
        return environment == .development
    }
    
    /// Enable network request/response logging
    static var enableNetworkLogging: Bool {
        return environment == .development
    }
    
    /// Enable crash reporting
    static var enableCrashReporting: Bool {
        return environment == .production || environment == .staging
    }
    
    /// Enable analytics
    static var enableAnalytics: Bool {
        return environment == .production || environment == .staging
    }
    
    /// Allow localhost connections (development only)
    static var allowLocalhostConnections: Bool {
        return environment == .development
    }
    
    // MARK: - Cache Configuration
    
    /// Cache duration in seconds
    static var cacheDuration: TimeInterval {
        switch environment {
        case .development:
            return 300 // 5 minutes for development
        case .staging:
            return 600 // 10 minutes for staging
        case .production:
            return 3600 // 1 hour for production
        }
    }
    
    /// Enable response caching
    static var enableCaching: Bool {
        return true // Enabled for all environments
    }
    
    // MARK: - Network Configuration
    
    /// Request timeout interval in seconds
    static var requestTimeout: TimeInterval {
        switch environment {
        case .development:
            return 30
        case .staging:
            return 30
        case .production:
            return 20
        }
    }
    
    /// Maximum number of retry attempts for failed requests
    static var maxRetryAttempts: Int {
        switch environment {
        case .development:
            return 2
        case .staging:
            return 3
        case .production:
            return 3
        }
    }
    
    // MARK: - App Information
    
    /// App version string
    static var appVersion: String {
        return Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
    }
    
    /// Build number
    static var buildNumber: String {
        return Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "1"
    }
    
    /// Bundle identifier
    static var bundleIdentifier: String {
        return Bundle.main.bundleIdentifier ?? "com.hestami.app"
    }
    
    /// Full version string (version + build)
    static var fullVersion: String {
        return "\(appVersion) (\(buildNumber))"
    }
    
    // MARK: - Debug Helpers
    
    /// Print current configuration (debug builds only)
    static func printConfiguration() {
        #if DEBUG
        AppLogger.app.debug("=== App Configuration ===")
        AppLogger.app.debug("Environment: \(environment.rawValue, privacy: .public)")
        AppLogger.app.debug("API Base URL: \(apiBaseURL, privacy: .public)")
        AppLogger.app.debug("Static Media Host: \(staticMediaHost, privacy: .public)")
        AppLogger.app.debug("App Version: \(fullVersion, privacy: .public)")
        AppLogger.app.debug("Verbose Logging: \(enableVerboseLogging, privacy: .public)")
        AppLogger.app.debug("Network Logging: \(enableNetworkLogging, privacy: .public)")
        AppLogger.app.debug("Cache Duration: \(cacheDuration, privacy: .public)s")
        AppLogger.app.debug("Request Timeout: \(requestTimeout, privacy: .public)s")
        AppLogger.app.debug("========================")
        #endif
    }
}

// MARK: - Configuration Validation

extension AppConfiguration {
    /// Validate configuration on app launch
    static func validate() {
        // Ensure URLs are valid
        guard URL(string: apiBaseURL) != nil else {
            AppLogger.critical("Invalid API base URL: \(apiBaseURL)")
            return
        }
        
        // Log configuration in debug builds
        #if DEBUG
        AppLogger.info("App configured for \(environment.rawValue) environment", category: AppLogger.app)
        printConfiguration()
        #endif
    }
}

File: frontend/mobile/google/android/MyHomeAgent/app/src/main/java/com/hestami_ai/myhomeagent/config/AppConfiguration.kt
package com.hestami_ai.myhomeagent.config

import android.content.Context
import android.content.pm.PackageManager
import com.hestami_ai.myhomeagent.BuildConfig
import timber.log.Timber

/**
 * Build environment types matching iOS implementation
 */
enum class BuildEnvironment {
    DEVELOPMENT,
    STAGING,
    PRODUCTION;

    companion object {
        val current: BuildEnvironment
            get() = if (BuildConfig.DEBUG) DEVELOPMENT else PRODUCTION
    }
}

/**
 * Centralized app configuration based on build environment.
 * Mirrors the iOS AppConfiguration.swift implementation.
 */
@Suppress("unused")
object AppConfiguration {

    // MARK: - Environment
    val environment: BuildEnvironment = BuildEnvironment.current

    // MARK: - API Configuration

    /** Base URL for API server */
    val apiBaseUrl: String
        get() = when (environment) {
            BuildEnvironment.DEVELOPMENT -> "https://dev-homeservices.hestami-ai.com"
            BuildEnvironment.STAGING -> "https://staging-homeservices.hestami-ai.com"
            BuildEnvironment.PRODUCTION -> "https://homeservices.hestami-ai.com"
        }

    /** Static media server host */
    val staticMediaHost: String
        get() = when (environment) {
            BuildEnvironment.DEVELOPMENT -> "dev-static.hestami-ai.com"
            BuildEnvironment.STAGING -> "staging-static.hestami-ai.com"
            BuildEnvironment.PRODUCTION -> "static.hestami-ai.com"
        }

    /** Static media server port */
    const val staticMediaPort: String = "443" // HTTPS port for all environments

    /** WebSocket base URL for chat (reserved for future use) */
    val wsBaseUrl: String
        get() = when (environment) {
            BuildEnvironment.DEVELOPMENT -> "wss://dev-homeservices.hestami-ai.com"
            BuildEnvironment.STAGING -> "wss://staging-homeservices.hestami-ai.com"
            BuildEnvironment.PRODUCTION -> "wss://homeservices.hestami-ai.com"
        }

    // MARK: - Feature Flags

    /** Enable verbose logging (debug builds only) */
    val enableVerboseLogging: Boolean
        get() = environment == BuildEnvironment.DEVELOPMENT

    /** Enable network request/response logging */
    val enableNetworkLogging: Boolean
        get() = environment == BuildEnvironment.DEVELOPMENT

    /** Enable crash reporting */
    val enableCrashReporting: Boolean
        get() = environment == BuildEnvironment.PRODUCTION || environment == BuildEnvironment.STAGING

    /** Enable analytics */
    val enableAnalytics: Boolean
        get() = environment == BuildEnvironment.PRODUCTION || environment == BuildEnvironment.STAGING

    /** Allow localhost connections (development only) */
    val allowLocalhostConnections: Boolean
        get() = environment == BuildEnvironment.DEVELOPMENT

    // MARK: - Cache Configuration

    /** Cache duration in seconds */
    val cacheDurationSeconds: Long
        get() = when (environment) {
            BuildEnvironment.DEVELOPMENT -> 300L // 5 minutes
            BuildEnvironment.STAGING -> 600L // 10 minutes
            BuildEnvironment.PRODUCTION -> 3600L // 1 hour
        }

    /** Enable response caching */
    const val enableCaching: Boolean = true

    // MARK: - Network Configuration

    /** Request timeout interval in seconds */
    val requestTimeoutSeconds: Long
        get() = when (environment) {
            BuildEnvironment.DEVELOPMENT -> 30L
            BuildEnvironment.STAGING -> 30L
            BuildEnvironment.PRODUCTION -> 20L
        }

    /** Maximum number of retry attempts for failed requests */
    val maxRetryAttempts: Int
        get() = when (environment) {
            BuildEnvironment.DEVELOPMENT -> 2
            BuildEnvironment.STAGING -> 3
            BuildEnvironment.PRODUCTION -> 3
        }

    // MARK: - App Information

    /** App version string */
    fun getAppVersion(context: Context): String {
        return try {
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            packageInfo.versionName ?: "1.0.0"
        } catch (_: PackageManager.NameNotFoundException) {
            "1.0.0"
        }
    }

    /** Build number / version code */
    fun getBuildNumber(context: Context): Long {
        return try {
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
                packageInfo.longVersionCode
            } else {
                @Suppress("DEPRECATION")
                packageInfo.versionCode.toLong()
            }
        } catch (_: PackageManager.NameNotFoundException) {
            1L
        }
    }

    /** Package name / bundle identifier */
    fun getPackageName(context: Context): String {
        return context.packageName
    }

    /** Full version string (version + build) */
    fun getFullVersion(context: Context): String {
        return "${getAppVersion(context)} (${getBuildNumber(context)})"
    }

    // MARK: - Debug Helpers

    /** Print current configuration (debug builds only) */
    fun printConfiguration(context: Context) {
        if (BuildConfig.DEBUG) {
            Timber.d("=== App Configuration ===")
            Timber.d("Environment: ${environment.name}")
            Timber.d("API Base URL: $apiBaseUrl")
            Timber.d("Static Media Host: $staticMediaHost")
            Timber.d("App Version: ${getFullVersion(context)}")
            Timber.d("Verbose Logging: $enableVerboseLogging")
            Timber.d("Network Logging: $enableNetworkLogging")
            Timber.d("Cache Duration: ${cacheDurationSeconds}s")
            Timber.d("Request Timeout: ${requestTimeoutSeconds}s")
            Timber.d("========================")
        }
    }

    /** Validate configuration on app launch */
    fun validate(context: Context): Boolean {
        // Ensure URLs are valid
        val isValid = try {
            java.net.URL(apiBaseUrl)
            true
        } catch (_: Exception) {
            Timber.e("Invalid API base URL: $apiBaseUrl")
            false
        }

        // Log configuration in debug builds
        if (BuildConfig.DEBUG) {
            Timber.i("App configured for ${environment.name} environment")
            printConfiguration(context)
        }

        return isValid
    }
}
